File: DynamicLayout.hs
-- DynamicLayout Haskell API — generate layout JSON from Haskell code.
--
-- Single-file module. Zero dependencies outside base (no lens, no aeson needed for generation).
-- Works on GHC 8.10+.
--
-- Example:
--   import DynamicLayout (begin, fieldset, label, endFieldset, button, endLayout)
--   main = putStrLn $ endLayout $
--     fieldset "Info" $ label "Hello from Haskell!" $
--     endFieldset $ button "ok" "OK" "primary" True $
--     begin "My Page"

module DynamicLayout where

import Data.List (intercalate)

data Builder = Builder { bbuf :: String, bdepth :: Int, bcomma :: Bool, bcnt :: Int }
  deriving Show

begin :: String -> Builder
begin t = Builder { bbuf = "{\"ui\":{\n    \"title\": \"" ++ esc t ++ "\",\n    \"layout\": [", bdepth = 2, bcomma = False, bcnt = 0 }

endLayout :: Builder -> String
endLayout b = bbuf b' ++ "\n  ]\n}"
  where b' = put "\n  ],\n  \"translations\": {},\n  \"userAccess\": { \"cancel\": true }" b

put :: String -> Builder -> Builder
put s b = b { bbuf = bbuf b ++ s }

indent :: Builder -> Builder
indent b = put ('\n' : replicate (bdepth b * 2) ' ') b

cm :: Builder -> Builder
cm b = if bcomma b then put "," b { bcomma = False } else b

esc :: String -> String
esc = concatMap (\c -> case c of
    '"'  -> "\\\""; '\\' -> "\\\\"; '\n' -> "\\n"; '\r' -> "\\r"; '\t' -> "\\t"
    _    -> [c])

kv :: String -> String -> Builder -> Builder
kv k v b = cm b' `put` "," `indent` `put` ('"' : esc k ++ "\": \"" ++ esc v ++ "\"")
  where b' = b { bcomma = True }

kvint :: String -> Int -> Builder -> Builder
kvint k v b = cm b' `put` "," `indent` `put` ('"' : esc k ++ "\": " ++ show v)
  where b' = b { bcomma = True }

kvbool :: String -> Builder -> Builder
kvbool k b = cm b' `put` "," `indent` `put` ('"' : esc k ++ "\": true")
  where b' = b { bcomma = True }

obj :: Builder -> Builder
obj b = cm b `put` "," `indent` `put` "{" `inc`
  where inc b0 = b0 { bdepth = bdepth b0 + 1, bcomma = False }

closeObj :: Builder -> Builder
closeObj b = b' { bdepth = bdepth b' - 1, bcomma = True } `indent` `put` "}"
  where b' = b { bcomma = False }

arr :: String -> Builder -> Builder
arr k b = cm b `put` "," `indent` `put` ('"' : esc k ++ "\": [") `inc`
  where inc b0 = b0 { bdepth = bdepth b0 + 1, bcomma = False }

closeArr :: Builder -> Builder
closeArr b = b' { bdepth = bdepth b' - 1, bcomma = True } `indent` `put` "]"
  where b' = b { bcomma = False }

key :: String -> Builder -> (String, Builder)
key p b = (p ++ "_" ++ show (bcnt b + 1), b { bcnt = bcnt b + 1 })

withKey :: (String -> Builder -> Builder) -> String -> Builder -> Builder
withKey f p b = let (k, b') = key p b in f k b'

-- Containers
fieldset :: String -> Builder -> Builder
fieldset t b = obj' $ kv "type" "FIELDSET" `kv "key" k `kv "title" t `arr "content"
  where (obj', (k, _)) = (obj, key "fs" b)

endFieldset :: Builder -> Builder
endFieldset = closeArr . closeObj

-- Display
label :: String -> Builder -> Builder
label t = withKey (\k -> obj . kv "type" "LABEL" . kv "key" k . kv "label" t . closeObj) "l"

alert :: String -> String -> Builder -> Builder
alert m c = withKey (\k -> obj . kv "type" "ALERT" . kv "key" k . kv "message" m . kv "color" c . closeObj) "a"

badge :: String -> String -> Builder -> Builder
badge t c = withKey (\k -> obj . kv "type" "BADGE" . kv "key" k . kv "title" t . kv "color" c . closeObj) "bd"

spacer :: Int -> Builder -> Builder
spacer px = withKey (\k -> obj . kv "type" "SPACER" . kv "key" k . kvint "width" px . closeObj) "sp"

-- Inputs
input :: String -> String -> Bool -> Builder -> Builder
input i l r = withKey (\k -> obj . kv "type" "INPUT" . kv "key" k . kv "id" i . kv "label" l . req . closeObj) "i"
  where req = if r then kvbool "required" else id

checkbox :: String -> String -> Builder -> Builder
checkbox i l = withKey (\k -> obj . kv "type" "CHECKBOX" . kv "key" k . kv "id" i . kv "label" l . closeObj) "cb"

-- Actions
button :: String -> String -> String -> Bool -> Builder -> Builder
button i t c d = withKey (\k -> obj . kv "type" "BUTTON" . kv "key" k . kv "id" i . kv "title" t . kv "color" c . def . closeObj) "btn"
  where def = if d then kvbool "default" else id

File: dynamiclayout.R
# DynamicLayout R API — generate layout JSON from R code.
#
# Single-file. Zero dependencies outside base R.
# Works on R 4.0+.
#
# Example:
#   source("dynamiclayout.R")
#   b <- dl_begin("My Page")
#   dl_fieldset(b, "Info")
#   dl_label(b, "Hello from R!")
#   dl_end_fieldset(b)
#   dl_button(b, "ok", "OK", "primary", TRUE)
#   cat(dl_end(b), "\n")

dl_put <- function(b, x) { b$buf <- paste0(b$buf, x); b }

dl_indent <- function(b) {
  b$buf <- paste0(b$buf, "\n", paste(rep("  ", b$depth), collapse=""))
  b
}

dl_cm <- function(b) {
  if (b$comma) b$buf <- paste0(b$buf, ",")
  b$comma <- FALSE; b
}

dl_esc <- function(b, s) {
  s <- gsub("\\\\", "\\\\\\\\", s)
  s <- gsub('"', '\\\\"', s)
  s <- gsub("\n", "\\\\n", s)
  b$buf <- paste0(b$buf, '"', s, '"')
  b
}

dl_kv <- function(b, key, val) {
  b <- dl_cm(b); b$buf <- paste0(b$buf, ","); b <- dl_indent(b)
  b <- dl_esc(b, key); b$buf <- paste0(b$buf, ": "); b <- dl_esc(b, val)
  b$comma <- TRUE; b
}

dl_kvint <- function(b, key, val) {
  b <- dl_cm(b); b$buf <- paste0(b$buf, ","); b <- dl_indent(b)
  b <- dl_esc(b, key); b$buf <- paste0(b$buf, ": ", val)
  b$comma <- TRUE; b
}

dl_kvbool <- function(b, key) {
  b <- dl_cm(b); b$buf <- paste0(b$buf, ","); b <- dl_indent(b)
  b <- dl_esc(b, key); b$buf <- paste0(b$buf, ": true")
  b$comma <- TRUE; b
}

dl_obj <- function(b) {
  b <- dl_cm(b); b$buf <- paste0(b$buf, ","); b <- dl_indent(b)
  b$buf <- paste0(b$buf, "{"); b$depth <- b$depth + 1; b$comma <- FALSE; b
}

dl_co <- function(b) {
  b$depth <- b$depth - 1; b <- dl_indent(b); b$buf <- paste0(b$buf, "}"); b$comma <- TRUE; b
}

dl_arr <- function(b, key) {
  b <- dl_cm(b); b$buf <- paste0(b$buf, ","); b <- dl_indent(b)
  b <- dl_esc(b, key); b$buf <- paste0(b$buf, ": [")
  b$depth <- b$depth + 1; b$comma <- FALSE; b
}

dl_ca <- function(b) {
  b$depth <- b$depth - 1; b <- dl_indent(b); b$buf <- paste0(b$buf, "]"); b$comma <- TRUE; b
}

dl_key <- function(b, prefix) {
  b$counter <- b$counter + 1
  paste0(prefix, "_", b$counter)
}

# Public API
dl_begin <- function(title) {
  b <- list(buf = "", depth = 0, comma = FALSE, counter = 0)
  b$buf <- '{"ui":{'
  b$depth <- 2; b$comma <- FALSE
  b <- dl_kv(b, "title", title)
  b <- dl_arr(b, "layout")
  b
}

dl_end <- function(b) {
  b <- dl_ca(b)
  b$comma <- FALSE; b$buf <- paste0(b$buf, ',')
  b$buf <- paste0(b$buf, '"translations":{},')
  b$buf <- paste0(b$buf, '"userAccess":{"cancel":true}')
  b$buf <- paste0(b$buf, '}}')
  b$buf
}

dl_fieldset <- function(b, title) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "FIELDSET"); b <- dl_kv(b, "key", dl_key(b, "fs"))
  b <- dl_kv(b, "title", title); b <- dl_arr(b, "content"); b
}
dl_end_fieldset <- function(b) { b <- dl_ca(b); b <- dl_co(b); b }

dl_row <- function(b) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "ROW"); b <- dl_kv(b, "key", dl_key(b, "r")); b <- dl_arr(b, "content"); b
}
dl_end_row <- function(b) { b <- dl_ca(b); b <- dl_co(b); b }

dl_label <- function(b, text) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "LABEL"); b <- dl_kv(b, "key", dl_key(b, "l"))
  b <- dl_kv(b, "label", text); b <- dl_co(b); b
}
dl_alert <- function(b, msg, color = "info") {
  b <- dl_obj(b); b <- dl_kv(b, "type", "ALERT"); b <- dl_kv(b, "key", dl_key(b, "a"))
  b <- dl_kv(b, "message", msg); b <- dl_kv(b, "color", color); b <- dl_co(b); b
}
dl_badge <- function(b, title, color = "primary") {
  b <- dl_obj(b); b <- dl_kv(b, "type", "BADGE"); b <- dl_kv(b, "key", dl_key(b, "bd"))
  b <- dl_kv(b, "title", title); b <- dl_kv(b, "color", color); b <- dl_co(b); b
}
dl_spacer <- function(b, px = 20) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "SPACER"); b <- dl_kv(b, "key", dl_key(b, "sp"))
  b <- dl_kvint(b, "width", px); b <- dl_co(b); b
}

dl_input <- function(b, id, label, required = FALSE) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "INPUT"); b <- dl_kv(b, "key", dl_key(b, "i"))
  b <- dl_kv(b, "id", id); b <- dl_kv(b, "label", label)
  if (required) b <- dl_kvbool(b, "required")
  b <- dl_co(b); b
}
dl_textarea <- function(b, id, label, rows = 3) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "TEXTAREA"); b <- dl_kv(b, "key", dl_key(b, "ta"))
  b <- dl_kv(b, "id", id); b <- dl_kv(b, "label", label)
  if (rows > 0) b <- dl_kvint(b, "rows", rows)
  b <- dl_co(b); b
}
dl_checkbox <- function(b, id, label) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "CHECKBOX"); b <- dl_kv(b, "key", dl_key(b, "cb"))
  b <- dl_kv(b, "id", id); b <- dl_kv(b, "label", label); b <- dl_co(b); b
}
dl_rating <- function(b, id, label) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "RATING"); b <- dl_kv(b, "key", dl_key(b, "rt"))
  b <- dl_kv(b, "id", id); b <- dl_kv(b, "label", label); b <- dl_co(b); b
}

dl_button <- function(b, id, title, color = "primary", is_default = FALSE) {
  b <- dl_obj(b); b <- dl_kv(b, "type", "BUTTON"); b <- dl_kv(b, "key", dl_key(b, "btn"))
  b <- dl_kv(b, "id", id); b <- dl_kv(b, "title", title); b <- dl_kv(b, "color", color)
  if (is_default) b <- dl_kvbool(b, "default")
  b <- dl_co(b); b
}

dl_section <- function(b, title, text) {
  b <- dl_fieldset(b, title); b <- dl_label(b, text); b <- dl_end_fieldset(b); b
}
